### PR TITLE
Add Futures Client and Socket

### PR DESCRIPTION
**Pull request by captholley**
Creates a new Client and Socket for Futures trading. Also adds in new objects for those classes that are different in the Futures API. Order of commits:

- [x] Futures Rest Client - Including Interface and necessary objects
- [x] Futures Socket Client - Including Interface and necessary stream objects
- [ ] Futures Rest Client Tests
- [ ] Futures Socket Client Tests
- [ ] Any additional items that may have been missed

### CODE REVIEW

**Comment by captholley** [comment]
I have now committed both the rest and socket clients. I wanted to get this sent through ASAP in case anyone was really pushing for this to finish up. I will try and complete the Tests over the next week or so.

File: Binance.Net/BinanceFuturesClient.cs [hunk 1]
```diff
@@ -0,0 +1,1452 @@
...
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Binance.Net.Interfaces;
+
+namespace Binance.Net
+{
+    /// <summary>
+    /// Client providing access to the Binance REST Api
```
Comment (JKorf) [inline Binance.Net/BinanceFuturesClient.cs]: Futures REST Api

File: Binance.Net/BinanceFuturesClient.cs [hunk 2]
```diff
@@ -0,0 +1,1452 @@
...
+            {
+                { "timestamp", GetTimestamp() }
+            };
+            parameters.AddOptionalParameter("recvWindow", receiveWindow?.ToString(CultureInfo.InvariantCulture) ?? defaultReceiveWindow.TotalMilliseconds.ToString(CultureInfo.InvariantCulture));
+
+            return await SendRequest<BinanceAccountInfo>(GetUrl(AccountInfoEndpoint, Api, SignedVersion), HttpMethod.Get, ct, parameters, true).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Requests to change the initial leverag of the given symbol
```
Comment (JKorf) [inline Binance.Net/BinanceFuturesClient.cs]: leverage

File: Binance.Net/BinanceFuturesSocketClient.cs [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+using CryptoExchange.Net.Authentication;
+using CryptoExchange.Net.Logging;
+using CryptoExchange.Net.Objects;
+using CryptoExchange.Net.Sockets;
+using Newtonsoft.Json.Linq;
+
+namespace Binance.Net
+{
+    /// <summary>
+    /// Client providing access to the Binance websocket Api
```
Comment (JKorf) [inline Binance.Net/BinanceFuturesSocketClient.cs]: futures websocket Api

File: Binance.Net/BinanceFuturesSocketClient.cs [hunk 4]
```diff
@@ -0,0 +1,744 @@
...
+    public class BinanceFuturesSocketClient : SocketClient, IBinanceFuturesSocketClient
+    {
+        #region fields
+        private static BinanceFuturesSocketClientOptions defaultOptions = new BinanceFuturesSocketClientOptions();
+        private static BinanceFuturesSocketClientOptions DefaultOptions => defaultOptions.Copy();
+
+        private readonly string baseCombinedAddress;
+
+        private const string AggregatedTradesStreamEndpoint = "@aggTrade";
+        private const string MarkPriceStreamEndpoint = "@markPrice";//TODO
```
Comment (JKorf) [inline Binance.Net/BinanceFuturesSocketClient.cs]: These TODO's still have to be implemented?

File: Binance.Net/BinanceFuturesSocketClient.cs [hunk 5]
```diff
@@ -0,0 +1,744 @@
...
+
+        /// <summary>
+        /// Create a new instance of BinanceSocketClient with default options
+        /// </summary>
+        public BinanceFuturesSocketClient() : this(DefaultOptions)
+        {
+        }
+
+        /// <summary>
+        /// Create a new instance of BinanceSocketClient using provided options
```
Comment (JKorf) [inline Binance.Net/BinanceFuturesSocketClient.cs]: BinanceFuturesSocketClient

File: Binance.Net/BinanceFuturesSocketClient.cs [hunk 6]
```diff
@@ -0,0 +1,744 @@
...
+        private const string OcoOrderUpdateEvent = "ORDER_TRADE_UPDATE";
+        private const string AccountPositionUpdateEvent = "outboundAccountPosition";
+        private const string BalanceUpdateEvent = "balanceUpdate";
+        private const string ListenKeyExpiredEvent = "listenKeyExpired";
+        #endregion
+
+        #region constructor/destructor
+
+        /// <summary>
+        /// Create a new instance of BinanceSocketClient with default options
```
Comment (JKorf) [inline Binance.Net/BinanceFuturesSocketClient.cs]: BinanceFuturesSocketClient

File: Binance.Net/BinanceFuturesSocketClient.cs [hunk 7]
```diff
@@ -0,0 +1,744 @@
...
+                        break;
+                    }
+                    case ListenKeyExpiredEvent:
+                    {
+                        log.Write(LogVerbosity.Debug, data);
+                        var result = Deserialize<BinanceStreamEvent>(token, false);
+                        if (result)
+                            onListenKeyExpired?.Invoke(result.Data);
+                        else
+                            log.Write(LogVerbosity.Warning, "Couldn't deserialize data received from account position stream: " + result.Error);
```
Comment (JKorf) [inline Binance.Net/BinanceFuturesSocketClient.cs]: warning messages are a bit scrambled

File: Binance.Net/Objects/BinanceFuturesClientOptions.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+        /// <summary>
+        /// The default receive window for requests
+        /// </summary>
+        public TimeSpan ReceiveWindow { get; set; } = TimeSpan.FromSeconds(5);
+
+        /// <summary>
+        /// ctor
+        /// </summary>
+        public BinanceFuturesClientOptions(): base("https://fapi.binance.com")
```
Comment (JKorf) [inline Binance.Net/Objects/BinanceFuturesClientOptions.cs]: Could you add an overload to specify a custom base address? Same for the socket client options

File: Binance.Net/Objects/BinanceFuturesPosition.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    public class BinanceFuturesPosition
+    {
+        /// <summary>
+        /// The entry price of the position
+        /// </summary>
+        public decimal EntryPrice { get; set; }
+        /// <summary>
+        /// Type of margin used for the position
+        /// </summary>
+        public FuturesMarginType MarginType { get; set; }
```
Comment (JKorf) [inline Binance.Net/Objects/BinanceFuturesPosition.cs]: Doesn't this need a specified converter?

File: Binance.Net/Objects/BinanceFuturesStreamLiquidation.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        public decimal AveragePrice { get; set; }
+        
+        /// <summary>
+        /// Liquidation Order Status
+        /// </summary>
+        [JsonProperty("X")]
+        public OrderStatus OrderStatus { get; set; }
+        
+        /// <summary>
+        /// Liquidation Last Fille Quantity
```
Comment (JKorf) [inline Binance.Net/Objects/BinanceFuturesStreamLiquidation.cs]: filled

File: Binance.Net/Objects/BinanceFuturesSymbol.cs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    public class BinanceFuturesSymbol
+    {
+        /// <summary>
+        /// Filters for order on this symbol
+        /// </summary>
+        public IEnumerable<BinanceSymbolFilter> Filters { get; set; } = new List<BinanceSymbolFilter>();
+        /// <summary>
+        /// The open margin percent
+        /// </summary>
+        public decimal MaintMarginPercent { get; set; }
```
Comment (JKorf) [inline Binance.Net/Objects/BinanceFuturesSymbol.cs]: Maint

File: Binance.Net/BinanceFuturesSocketClient.cs (hunk 4 above)
Comment (captholley) [reply Binance.Net/BinanceFuturesSocketClient.cs]: I had put this down as a note to myself not to forget. These are complete now

File: Binance.Net/Objects/BinanceFuturesPosition.cs (hunk 9 above)
Comment (captholley) [reply Binance.Net/Objects/BinanceFuturesPosition.cs]: I added the converters with the first commit. Sorry for not being clear with that.

https://github.com/JKorf/Binance.Net/pull/279/commits/b72a675c02f96dc9f36d73e15c9ed86e9ade3e4f#diff-53cef88717420750dd95c29891fe67bc

**Comment by captholley** [comment]
> Thanks for the work you put into this, it looks good!
> I just have some small comments, most of them some typo in comments or something.
> 
> I assume you tested it yourself?

I have fixed the issues you posted and pushed a new commit. I have not done the unit testing yet, but I have tested a few of the methods running with my own program. I found that the stream information I actually received didn't always match the information provided in the documentation. I also noticed that the user stream did not include some of the items included in the spot trading user stream, but Binance is moving fast and I am assuming they will add these features in the future so I left the cases that might not be used as of right now (Execution, Account Balance, Account position).

**Comment by nikkozp** [comment]
@captholley thx for PR
@JKorf  can you add this PR to the alpha version on NuGet or to a separate branch? Now I'm starting my large-scale project and will thoroughly test everything, and help with new PR :)

File: Binance.Net/Objects/BinanceFuturesSymbol.cs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        /// <summary>
+        /// The status of the symbol
+        /// </summary>
+        [JsonConverter(typeof(SymbolStatusConverter))]
+        public SymbolStatus Status { get; set; }
+        /// <summary>
+        /// Allowed order types
+        /// </summary>
+        [JsonProperty(ItemConverterType = typeof(OrderTypeConverter))]
+        public IEnumerable<OrderType> OrderTypes { get; set; } = new List<OrderType>();
```
Comment (nikkozp) [inline Binance.Net/Objects/BinanceFuturesSymbol.cs]: TRAILING_STOP_MARKET

**Comment by spbsmile** [comment]
alpha version on NuGet. plz)

**Comment by JKorf** [comment]
I've merged this PR into a new branch -> alpha-futures. I've also pushed an alpha version to Nuget so you guys can start using it.

If/When creating additional PRs for the futures API please use the alpha-futures branch. 
Thanks for your work on this!

**Pull request closed without merging**